als/src/CuMFJNIInterface.cpp

```diff
@@ -24,107 +24,108 @@
+				memcpy(thetaTHost + index*f, factorfloat, sizeof(float)*f);
+				index ++;
+				env->ReleasePrimitiveArrayCritical((jfloatArray)factor, factorfloat, 0);
+				env->DeleteLocalRef(factor);
+			}
+			env->DeleteLocalRef(factorsPerBlock);
+		}
+		// get a pointer to the raw input data, pinning them in memory
+		csrRowIndexHostPtr = (jint*) env->GetPrimitiveArrayCritical(csrRow, 0);
+		csrColIndexHostPtr = (jint*) env->GetPrimitiveArrayCritical(csrCol, 0);
+		csrValHostPtr = (jfloat*) env->GetPrimitiveArrayCritical(csrVal, 0);
 
-	/*
-	printf("thetaTHost of len %d: \n", n*f);
-    for (int i = 0; i < n*f; i++) {
-	  printf("%f ", thetaTHost[i]);
-    }
-	printf("\n");
-	*/
-	int * d_csrRowIndex = 0;
-	int * d_csrColIndex = 0;
-	float * d_csrVal = 0;
+		/*
+		printf("thetaTHost of len %d: \n", n*f);
+		for (int i = 0; i < n*f; i++) {
+		  printf("%f ", thetaTHost[i]);
+		}
+		printf("\n");
+		*/
+		int * d_csrRowIndex = 0;
+		int * d_csrColIndex = 0;
+		float * d_csrVal = 0;
 
-	cudacall(cudaMalloc((void** ) &d_csrRowIndex,(m + 1) * sizeof(float)));
-	cudacall(cudaMalloc((void** ) &d_csrColIndex, nnz * sizeof(float)));
-	cudacall(cudaMalloc((void** ) &d_csrVal, nnz * sizeof(float)));
-	cudacall(cudaMemcpyAsync(d_csrRowIndex, csrRowIndexHostPtr,(size_t ) ((m + 1) * sizeof(float)), cudaMemcpyHostToDevice, cuda_stream));
-	cudacall(cudaMemcpyAsync(d_csrColIndex, csrColIndexHostPtr,(size_t ) (nnz * sizeof(float)), cudaMemcpyHostToDevice, cuda_stream));
-	cudacall(cudaMemcpyAsync(d_csrVal, csrValHostPtr,(size_t ) (nnz * sizeof(float)),cudaMemcpyHostToDevice, cuda_stream));
-	cudaStreamSynchronize(cuda_stream);
+		cudacall(cudaMalloc((void** ) &d_csrRowIndex,(m + 1) * sizeof(float)));
+		cudacall(cudaMalloc((void** ) &d_csrColIndex, nnz * sizeof(float)));
+		cudacall(cudaMalloc((void** ) &d_csrVal, nnz * sizeof(float)));
+		cudacall(cudaMemcpyAsync(d_csrRowIndex, csrRowIndexHostPtr,(size_t ) ((m + 1) * sizeof(float)), cudaMemcpyHostToDevice, cuda_stream));
+		cudacall(cudaMemcpyAsync(d_csrColIndex, csrColIndexHostPtr,(size_t ) (nnz * sizeof(float)), cudaMemcpyHostToDevice, cuda_stream));
+		cudacall(cudaMemcpyAsync(d_csrVal, csrValHostPtr,(size_t ) (nnz * sizeof(float)),cudaMemcpyHostToDevice, cuda_stream));
+		cudaStreamSynchronize(cuda_stream);
 
-	// un-pin the host arrays, as we're done with them
-	env->ReleasePrimitiveArrayCritical(csrRow, csrRowIndexHostPtr, 0);
-	env->ReleasePrimitiveArrayCritical(csrCol, csrColIndexHostPtr, 0);
-	env->ReleasePrimitiveArrayCritical(csrVal, csrValHostPtr, 0);
+		// un-pin the host arrays, as we're done with them
+		env->ReleasePrimitiveArrayCritical(csrRow, csrRowIndexHostPtr, 0);
+		env->ReleasePrimitiveArrayCritical(csrCol, csrColIndexHostPtr, 0);
+		env->ReleasePrimitiveArrayCritical(csrVal, csrValHostPtr, 0);
 
-	printf("\tdoALSWithCSR with m=%d,n=%d,f=%d,nnz=%d,lambda=%f \n.", m, n, f, nnz, lambda);
-	try{
+		printf("\tdoALSWithCSR with m=%d,n=%d,f=%d,nnz=%d,lambda=%f \n.", m, n, f, nnz, lambda);

```

Comment:
It would be nice if the user could disable the printf statements.  One possibility would be to surround them with a #ifdef, and use multiple makefile targets, so that the default target doesn't print, but something like "make debug" does print.